src/binary-writer-spec.h

```diff
@@ -17,21 +17,43 @@
 #ifndef WABT_BINARY_WRITER_SPEC_H_
 #define WABT_BINARY_WRITER_SPEC_H_
 
+#include <functional>
+#include <utility>
+#include <vector>
+
 #include "src/binary-writer.h"
 #include "src/common.h"
 #include "src/ir.h"
 
 namespace wabt {
 
-struct WriteBinarySpecOptions {
-  Stream* log_stream = nullptr;
-  const char* json_filename = nullptr;
-  WriteBinaryOptions write_binary_options;
+struct FilenameMemoryStreamPair {
+  FilenameMemoryStreamPair(string_view filename,
+                           std::unique_ptr<MemoryStream> stream)
+      : filename(filename), stream(std::move(stream)) {}
+  std::string filename;
+  std::unique_ptr<MemoryStream> stream;
 };
 
-Result WriteBinarySpecScript(struct Script*,
-                             const char* source_filename,
-                             const WriteBinarySpecOptions*);
+typedef std::function<Stream*(string_view filename)>
+    WriteBinarySpecStreamFactory;
+
+Result WriteBinarySpecScript(Stream* json_stream,
+                             WriteBinarySpecStreamFactory module_stream_factory,
+                             Script*,
+                             string_view source_filename,
+                             string_view module_filename_noext,
+                             const WriteBinaryOptions*);
```

Comment:
What are some args not named?  Is there some kind of rule here?

Reply:
I often omit names that are redundant, kind of like Blink style. So you don't have `Script* script` or `const WriteBinaryOptions* options`. But it's not really anything formalized.